service/src/main/kotlin/app/cash/backfila/dashboard/CloneBackfillAction.kt

```diff
@@ -0,0 +1,217 @@
+              request.dry_run
+          )
+      )
+    } catch (e: Exception) {
+      logger.info(e) { "PrepareBackfill on `${dbData.serviceName}` failed" }
+      throw BadRequestException("PrepareBackfill on `${dbData.serviceName}` failed: ${e.message}", e)
+    }
+    val partitions = prepareBackfillResponse.partitions
+    if (partitions.isEmpty()) {
+      throw BadRequestException("PrepareBackfill returned no partitions")
+    }
+    if (partitions.any { it.partition_name == null }) {
+      throw BadRequestException("PrepareBackfill returned unnamed partitions")
+    }
+    if (partitions.distinctBy { it.partition_name }.size != partitions.size) {
+      throw BadRequestException("PrepareBackfill did not return distinct partition names:" +
+          " ${partitions.map { it.partition_name }}")
+    }
+
+    val combinedParams = request.parameter_map.plus(prepareBackfillResponse.parameters)
+
+    val backfillRunId = transacter.transaction { session ->
+      val backfillRun = DbBackfillRun(
+          dbData.serviceId,
+          dbData.registeredBackfillId,
+          combinedParams,
+          BackfillState.PAUSED,
+          caller.get()?.user,
+          request.scan_size,
+          request.batch_size,
+          request.num_threads,
+          request.backoff_schedule,
+          request.dry_run,
+          request.extra_sleep_ms
+      )
+      session.save(backfillRun)
+
+      if (request.range_clone_type == RangeCloneType.NEW) {
+        for (partition in partitions) {
+          val dbRunPartition = DbRunPartition(
+              backfillRun.id,
+              partition.partition_name,
+              partition.backfill_range ?: KeyRange.Builder().build(),
+              backfillRun.state,
+              partition.estimated_record_count
+          )
+          session.save(dbRunPartition)
+        }
+      } else {
+        // Verify partitions match source backfill, use per partition ranges.
+        val sourceBackfill = session.load<DbBackfillRun>(Id(id))
+        // Source partitions have to match new partitions.
+        val sourcePartitions = sourceBackfill.partitions(session, queryFactory)
+        if (partitions.map { it.partition_name }.toSet() != sourcePartitions.map { it.partition_name }.toSet()) {
+          throw BadRequestException("Can't clone backfill ranges from `$id`, newly computed partitions don't match.")
+        }
+
+        for (sourcePartition in sourcePartitions) {
+          val dbRunPartition = DbRunPartition(
+              backfillRun.id,
+              sourcePartition.partition_name,
+              sourcePartition.backfillRange(),
+              backfillRun.state,
+              sourcePartition.estimated_record_count
+          )
+          // Copy the cursor if continuing, otherwise just leave blank to start from beginning.
+          if (request.range_clone_type == RangeCloneType.CONTINUE) {
+            dbRunPartition.pkey_cursor = sourcePartition.pkey_cursor
```

Comment:
If its null does it still use the start range? There should probably be a test for this to prove this.

Reply:
null = start from the beginning

when null cursor is passed, the operator figures out the cursor from the range start